spec.bs

```diff
@@ -1686,16 +1683,26 @@ To <dfn>convert a modifier to a string</dfn> given a [=part/modifier=] |modifier
   1. Set |result|["{{URLPatternInit/hash}}"] to |hash|.
   1. Let |baseURL| be null.
   1. If |init|["{{URLPatternInit/baseURL}}"] is not null:
+    <div class="note">
+      The base URL can be used to supply additional context, but for each component, if |init| includes a component which is at least as specific as one in the base URL, none is inherited.
+
+      A component is more specific if it appears later in one of the following two lists (which are very similar to the order they appear in the URL syntax):
+
+      * protocol, hostname, port, pathname, search, hash
+      * protocol, hostname, port, username, password
+
+      Username and password are also never inherited when constructing patterns, only for a URL being matched.
+    </div>
     1. Set |baseURL| to the result of [=URL parser|parsing=] |init|["{{URLPatternInit/baseURL}}"].
     1. If |baseURL| is failure, then throw a {{TypeError}}.
-    1. Set |result|["{{URLPatternInit/protocol}}"] to the result of [=processing a base URL string=] given |baseURL|'s [=url/scheme=] and |type|.
-    1. Set |result|["{{URLPatternInit/username}}"] to the result of [=processing a base URL string=] given |baseURL|'s [=url/username=] and |type|.
-    1. Set |result|["{{URLPatternInit/password}}"] to the result of [=processing a base URL string=] given |baseURL|'s [=url/password=] and |type|.
-    1. Set |result|["{{URLPatternInit/hostname}}"] to the result of [=processing a base URL string=] given |baseURL|'s [=url/host=] and |type|.
-    1. Set |result|["{{URLPatternInit/port}}"] to the result of [=processing a base URL string=] given |baseURL|'s [=url/port=] and |type|.
-    1. Set |result|["{{URLPatternInit/pathname}}"] to the result of [=processing a base URL string=] given the result of [=URL path serializing=] |baseURL| and |type|.
-    1. Set |result|["{{URLPatternInit/search}}"] to the result of [=processing a base URL string=] given |baseURL|'s [=url/query=] and |type|.
-    1. Set |result|["{{URLPatternInit/hash}}"] to the result of [=processing a base URL string=] given |baseURL|'s [=url/fragment=] and |type|.
+    1. If |init|["{{URLPatternInit/protocol}}"] does not [=map/exist=], then set |result|["{{URLPatternInit/protocol}}"] to the result of [=processing a base URL string=] given |baseURL|'s [=url/scheme=] and |type|.
+    1. If |type| is not "`pattern`" and |init| [=map/contains=] none of "{{URLPatternInit/protocol}}", "{{URLPatternInit/hostname}}", "{{URLPatternInit/port}}" and "{{URLPatternInit/username}}", then set |result|["{{URLPatternInit/username}}"] to the result of [=processing a base URL string=] given |baseURL|'s [=url/username=] and |type|.
```

Comment:
and -> or here and below, I think?

Reply:
I'm unsure. The WHATWG style guide is silent and I haven't found another style reference, probably because "none of" in formal writing generally doesn't go directly into a list. To me "none of A, B, and C" seems most natural (i.e., none of the elements in this set, which consists of A, B, and C), but "or" doesn't sound wrong either.

Another alternative would be "neither...nor", but for three or more items it's usual to repeat "nor" which gets pretty tedious for like 7 items that we get to (even more than the famous [USPS creed](https://en.wikipedia.org/wiki/United_States_Postal_Service_creed)):
> If init contains neither "protocol", nor "hostname", nor "port", nor "pathname", nor "search", nor "hash", ...

Yet another would be to use "any" instead of "none" and move the negation earlier, but it reads as a little more verbose too:
> If init does not contain any of "protocol", "hostname", "port", "pathname", "search", and "hash", ...

I don't think it's ambiguous in any case; what do you think?

Reply:
ChatGPT agrees with "and": https://chat.openai.com/share/550e4e1f-39cf-4a01-8664-85cdbb399d2b. (Bard is incoherent.)

I might go with

> If init does not contain "protocol", "hostname", "port", "pathname", "search", or "hash", ...

but I don't think we need to spend any more time on this. I agree people will understand either way :)